### PR TITLE
feat: extends stamps capacity

### LINKED ISSUE #311: Add support for extending existing postage stamp capacity
Related and similar to #310

Basically instead of buying a new postage stamp every time the usage threshold is reached with this feature it would be possible to extend the capacity (`depth`) of the postage stamp using the [dilute](https://docs.ethswarm.org/api/#tag/Postage-Stamps/paths/~1stamps~1dilute~1%7Bid%7D~1%7Bdepth%7D/patch) endpoint.

### PR DESCRIPTION
**Pull request by luissanchez0305**
Closes #311 

Basically, instead of buying a new postage stamp every time the usage threshold is reached with this feature, it would be possible to extend the capacity (depth) of the postage stamp using the [dilute](https://docs.ethswarm.org/api/#tag/Postage-Stamps/paths/~1stamps~1dilute~1%7Bid%7D~1%7Bdepth%7D/patch) endpoint.

### CODE REVIEW

File: src/stamps.ts [hunk 1]
```diff
@@ -300,6 +315,30 @@ export class StampsManager {
...
+
+    const { usageThreshold } = config
+
+    // Get all used stamps sorted by usage from most used to least
+    this.usableStamps = filterUsableStampsExtendsCapacity(stamps, usageThreshold)
+
+    for (const stamp of this.usableStamps) {
+      try {
+        logger.debug(`extending stamp capacity: ${stamp.batchID}`)
+        await topUpStamp(beeDebug, stamp.batchID, (Number(stamp.amount) * 2).toString())
```
Comment (AuHau) [inline src/stamps.ts:332]: Oooooh yeah, this is a problem 🙈 The reason why the `amount` is a string and not a number is that we are calculating with very big amounts which are hitting the limits of the JS number engine (see for example [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER)). Sooo, unfortunately, you can't just convert it to a number and do the math as there is a potential loss of precision in the game, which generally is a bad idea when you are working with money ;-)

I would suggest using `BigInt` which is native in JS (in Node it is available since 10.x so it is safe to use now as we support >=14.x), please see [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) for more details on how to use it.

File: src/stamps.ts [hunk 2]
```diff
@@ -111,6 +111,24 @@ export function filterUsableStampsExtends(stamps: PostageBatch[]): PostageBatch[
...
+/**
+ * Filter the stamps and only return those that are usable and sort by usage in a increasing order
+ *
+ * @param stamps Postage stamps to be filtered
+ *
+ * @returns Filtered stamps soltered by usage
+ */
+export function filterUsableStampsExtendsCapacity(stamps: PostageBatch[], usageThreshold: number): PostageBatch[] {
+  const usableStamps = stamps
+    // filter to get stamps that have the right depth, amount and are not fully used or expired
```
Comment (AuHau) [inline src/stamps.ts:123]: Outdated comment, please fix.

File: src/stamps.ts [hunk 3]
```diff
@@ -148,6 +166,10 @@ export async function topUpStamp(beeDebug: BeeDebug, postageBatchId: string, amo
   return stamp
 }
 
+export async function diluteStamp(beeDebug: BeeDebug, postageBatchId: string, depth: number) {
```
Comment (AuHau) [inline src/stamps.ts:169]: I would argue that this function is not needed. You can call the `diluteBatch` on `beeDebug` directly...

**Review by AuHau — CHANGES_REQUESTED** [review]
Few general things we need to address before moving forward.

Firstly, you are treating this feature as another exclusive mode of operation of gateway-proxy's stamp management. I don't think that is correct though. The whole gateway-proxy layout of features is getting quite complicated and once Attila is back we really should think of some better architecture for this. But in meanwhile...

What I see is that there is sort of two modes of the postage management. Either it would make sense to run it with `autobuy` enabled, which basically allows uploading content to Swarm that has a limited life-span on the network and the extending modes then does not really make sense (at least to me, feel free to come up with use-case proving me wrong). Or, you want to make sure that the uploaded stays in some form on the network, for which there are the extending stamp's features.

So, basically, I can see the possibility that when stamps management is not running in `hardcoded` or `autobuy` it can run with both `extendsTTL` and `extendsCapacity` modes... I think we should support this use case.

Secondly, we have quite neglected testing in your previous PRs. We should fix that later on, but lets start with this PR first. Vojtech created quite a nice tooling for testing of the StampsManager (eq. you have stamps DB, you can set their values etc.), so please use it in order to test thoroughly the (edge) cases of the `extendsCapacity` mode. Also, there should be maybe some tests that cover the mentioned interoperability of the `extendsCapacity` and `extendsTtl` modes. Lastly, please structure the tests so that each mode has its own `describe` section.

File: src/config.ts [hunk 4]
```diff
@@ -137,14 +150,43 @@ export function getStampsConfig({
...
+      POSTAGE_TTL_MIN,
+      POSTAGE_EXTENDSTTL,
+      POSTAGE_EXTENDS_CAPACITY,
+      refreshPeriod,
+      beeDebugApiUrl,
+    )
+  }
+}
+
+export function validateStampPostageConfig(
```
[the remark this answers is not in the document]
Comment (luissanchez0305) [reply src/config.ts:177]: I was under the impression that complexity resolution would be the least code possible, but this refactor did the trick and put the complexity below 19. So much to learn

File: README.md [hunk 5]
```diff
@@ -112,6 +114,14 @@ export BEE_DEBUG_API_URL=http://localhost:1635
 npm run start
 ```
 
+#### 5. Extends stamps capacity
+
+```sh
+export POSTAGE_EXTENDS_CAPACITY=true
```
Comment (Cafe137) [inline README.md:120]: Also document `POSTAGE_USAGE_THRESHOLD` here please, if I have seen correctly it determines when the stamp is diluted

File: README.md [hunk 6]
```diff
@@ -130,26 +140,28 @@ npm run start
...
+| AUTH_SECRET              | undefined                   | Authentication secret, disabled if not set (this secret is checked in the request header `authorization`). |
+| HOSTNAME                 | localhost                   | Hostname of the proxy. Required for Bzz.link support.                                                      |
+| PORT                     | 3000                        | Port of the proxy.                                                                                         |
+| POSTAGE_STAMP            | undefined                   | Postage stamp that should be used for all upload requests. If provided, the autobuy feature is disabled.   |
+| POSTAGE_DEPTH            | undefined                   | Postage stamp depth to be used when buying new stamps or selecting existing stamps.                        |
+| POSTAGE_AMOUNT           | undefined                   | Postage stamp amount to be used when buying new stamps or selecting existing stamps.                       |
+| POSTAGE_USAGE_THRESHOLD  | 0.7                         | Usage percentage at which new postage stamp will be bought (value between 0 and 1).                        |
```
Comment (Cafe137) [inline README.md:153]: Also document its role in dilute mode please

File: test/stamps.spec.ts [hunk 7]
```diff
@@ -241,3 +241,16 @@ describe('topUpStamp', () => {
...
   })
 })
+
+describe('extendsCapacity', () => {
+  it('should extend stamps capacity', async () => {
+    const beeDebug = new BeeDebug(url)
+
+    const stamp = await buyNewStamp(defaultDepth, defaultAmount, beeDebug)
+
+    await topUpStamp(beeDebug, stamp.batchId, (Number(stamp.stamp.amount) * 2).toString())
```
Comment (Cafe137) [inline test/stamps.spec.ts:251]: Hmmm, maybe @agazso or @AuHau can confirm it, but IIRC `topup` is additive instead of specifying the target `amount`. So the `* 2` is unnecessary here.

Comment (Cafe137) [reply test/stamps.spec.ts:251]: Also as a general comment (maybe concerns @agazso the most), we would need an atomic topup + dilute operation on blockchain level in order to make this feature solid, otherwise one of these operations can fail (http network issue, gas issue, non-graceful shutdown, running out of funds) making stamps grow asymmetrically.

Comment (Cafe137) [reply test/stamps.spec.ts:251]: Oops, meant this comments for `src/stamps.ts`, not the test 🙂

**Comment by Cafe137** [comment]
I am running with this env:

```
BEE_DEBUG_API_URL: 'http://localhost:1635',
POSTAGE_EXTENDS_CAPACITY: 'true',
}
```

Have a clean Bee, no stamps.

And it tries to buy a new stamp:

```
time="2022-10-14T11:43:02.100Z" level="info" msg="buying new stamp"
```

Which fails due to missing envs

```
time="2022-10-14T11:43:02.107Z" level="error" msg="failed to refresh on extends postage stamps Value has to be bigger then zero" value="0"
at ./src/stamps.ts:150:34
at ./src/stamps.ts:8:71"
```

But I think it should not buy stamps at all

File: src/stamps.ts [hunk 8]
```diff
@@ -236,45 +256,56 @@ export class StampsManager {
...
           await this.verifyUsableStamps(beeDebug, ttlMin, config, amount)
         }
       }
+
+      for (const stamp of this.usableStampsExtendsCapacity) {
+        try {
+          logger.debug(`extending stamp capacity: ${stamp.batchID}`)
+          await topUpStamp(beeDebug, stamp.batchID, (BigInt(stamp.amount) * BigInt(2)).toString())
+          await beeDebug.diluteBatch(stamp.batchID, stamp.depth + 1)
+          logger.info(`capacity extended for stamp ${stamp.batchID}`)
```
Comment (Cafe137) [inline src/stamps.ts:286]: In the extendsTTL feature, after the topup operation there is a 1 minute wait:

https://github.com/ethersphere/gateway-proxy/blob/master/src/stamps.ts#L280

Topup and dilute are the same in this regard, they both require a short while for the confirmation before Bee accepts and updates to the new values.

So if we have a wait on purpose in the other mode, let's have it here as well, or I am fine not having it in either places if it is not important, but keep it consistent otherwise it is confusing why it only exists for one of the modes.

**Comment by Cafe137** [comment]
Also another observation when `POSTAGE_REFRESH_PERIOD` is low

`EXTENDSTTL` does not run `topup` on a stamp again and again, instead it waits for the existing operation to complete.

`EXTENDS_CAPACITY`, however runs `topup` and `dilute` in every cycle, even if there is already an operation in progress.

File: README.md (hunk 5 above)
Comment (luissanchez0305) [reply README.md:120]: No, it doesn't determine _when_, it determine _how much_ the stamp have been use (aka the usage level of the stamp) for it to be extended. It uses the DEFAULT_POSTAGE_USAGE_THRESHOLD variable, so it is optional. I'll specify this on the **Usage** section of the README.md

File: README.md (hunk 6 above)
Comment (luissanchez0305) [reply README.md:153]: I don't use `POSTAGE_USAGE_THRESHOLD` to dilute the stamp. I use the stamp amount and depth to extend its capacity. Can you elaborate more on this, please?

File: test/stamps.spec.ts (hunk 7 above)
Comment (luissanchez0305) [reply test/stamps.spec.ts:251]: I'll wait for @AuHau @agazso 's insight on this

File: README.md (hunk 6 above)
Comment (Cafe137) [reply README.md:153]:
```
filterUsableStampsExtendsCapacity(stamps: PostageBatch[], usageThreshold: number): PostageBatch[]
```

This function uses `usageThreshold`, which is made from `Number(POSTAGE_USAGE_THRESHOLD || DEFAULT_POSTAGE_USAGE_THRESHOLD)`

File: README.md (hunk 5 above)
Comment (Cafe137) [reply README.md:120]:
```
filterUsableStampsExtendsCapacity(stamps: PostageBatch[], usageThreshold: number): PostageBatch[]
```

This function uses `usageThreshold`, which is made from `Number(POSTAGE_USAGE_THRESHOLD || DEFAULT_POSTAGE_USAGE_THRESHOLD)`

File: README.md (hunk 6 above)
Comment (luissanchez0305) [reply README.md:153]: Oh I see, yes the `usageThreshold` variable is used to get the list of stamps that have a _usage_ more extensive than that threshold value and it need to be documented on the README according to the feature that it's been use

File: test/stamps.spec.ts (hunk 7 above)
Comment (luissanchez0305) [reply test/stamps.spec.ts:251]: @Cafe137 I already test the additive property of topup call, and yes, no need to double up the amount

File: src/stamps.ts [hunk 9]
```diff
@@ -236,65 +260,88 @@ export class StampsManager {
...
     }
   }
 
-  async verifyUsableStamps(
-    beeDebug: BeeDebug,
-    ttlMin: number,
-    config: StampsConfigAutobuy | StampsConfigExtends,
-    amount: string,
-  ) {
-    for (let i = 0; i < this.usableStamps!.length; i++) {
```
Comment (AuHau) [inline src/stamps.ts:270]: The fact that you have replaced the `this.usableStamps` for `this.usableStampsExtendsTTL` makes me think that you have broken the `autobuy` mode with it. Isn't it true?

File: test/stamps.spec.ts (hunk 7 above)
Comment (AuHau) [reply test/stamps.spec.ts:251]: Oh, good catch @Cafe137 about the additive side!

And yes, agree about the atomicity problem here, I am afraid though that with the current Bee API this is not possible for us to handle :-( https://github.com/ethersphere/bee/issues/3359 would help us with this.

**Review by AuHau — CHANGES_REQUESTED** [review]
I am sorry, but the code gets too messy beyond what I can understand and reason about.  **But** the fact that we have cleared out how the Stamps Manager can operate (eq. that autobuy and extends functionality are exclusive), gives me an idea of how to improve the situation. Let's discuss it today on the call later on.

There are several other smaller issues I have identified. One of the major ones is regarding the `isNaN` check and generally input validation, which we should tackle (as per #393 issue).

Then there is the testing. I am sorry but the tests you have are very non-sufficient, we will have to improve that to ensure that the higher complexity that this project grows is working correctly. Lets discuss it on the call as well so we can get aligned with what I expect here.

**Comment by AuHau** [comment]
Ad @Cafe137 [comment](https://github.com/ethersphere/gateway-proxy/pull/361#issuecomment-1278896421):

>  But I think it should not buy stamps at all

I don't agree with that. IMHO it should buy one stamp and then keep it extending...

File: src/stamps.ts (hunk 9 above)
Comment (luissanchez0305) [reply src/stamps.ts:270]: No, this was resolved by using a separate one for autobuy. But it was completely modified by the refactor
